### PR TITLE
Add --parameters to frida-inject to allow specify parameters using a JSON string similar to frida-gadget

### PR DESCRIPTION
**Pull request by eugenekolo**
My first Vala code, so it might not be vala-thonic

`./frida-inject -p 16920 --script test.js --parameters '{"paramBool": true, "paramArray": [1,2,3], "paramInt": 1, "paramStr": "str", "paramObj": {"a": ["b", "c", {}]}}'`

```
stage early
parameters {"paramBool":true,"paramArray":[1,2,3],"paramInt":1,"paramStr":"str","paramObj":{"a":["b","c",{}]}}
```

### CODE REVIEW

**Comment by eugenekolo** [comment]
Done. Moved to where the other args are passed.

Tested now with passing blank parameters too. Same as Gadget handling, where it passes a {}.

Testing:
```
kiwi:/data/local/tmp # cat test.js 
'use strict';

rpc.exports = {
    init: function (stage, parameters) {
	console.log('stage', stage);
	console.log('parameters', JSON.stringify(parameters));
    },
};

kiwi:/data/local/tmp # ./frida-inject -p 1256 --script test.js --parameters '{"parmBool": true, "paramArray": [1,2,3], "paramInt": 1, "paramStr": "str", "paramObj": {"a": ["b", "c", {}]}}'                                                    <
stage early
parameters {"paramBool":true,"paramArray":[1,2,3],"paramInt":1,"paramStr":"str","paramObj":{"a":["b","c",{}]}}

kiwi:/data/local/tmp # ./frida-inject -p 1256 --script test.js --parameters ''                                                                                 
stage early
parameters {}

kiwi:/data/local/tmp # ./frida-inject -p 1256 --script test.js --parameters '{}'                                                                               
stage early
parameters {}

kiwi:/data/local/tmp # ./frida-inject -p 1256 --script test.js                                                                                                 
stage early
parameters {}
```

File: inject/inject.vala [hunk 1]
```diff
@@ -59,7 +61,18 @@ namespace Frida.Inject {
 			script_source = read_stdin ();
 		}
 
-		application = new Application (target_pid, target_name, script_path, script_source, enable_jit, enable_development);
+		Json.Node parameters = new Json.Node (Json.NodeType.OBJECT);
```
Comment (oleavr) [inline inject/inject.vala]:
```suggestion
		var parameters = new Json.Node (Json.NodeType.OBJECT);
```

File: inject/inject.vala [hunk 2]
```diff
@@ -59,7 +61,18 @@ namespace Frida.Inject {
...
+		parameters.set_object (new Json.Object ()); 
+		if (parameters_str != null && parameters_str != "") {
```
Comment (oleavr) [inline inject/inject.vala]: The empty string case should be an error. We can just ignore that here and let the parsing below fail.

File: inject/inject.vala [hunk 3]
```diff
@@ -59,7 +61,18 @@ namespace Frida.Inject {
...
+			try {
+				var parser = new Json.Parser ();
+				parser.load_from_data (parameters_str, -1);
+				parameters.set_object (parser.get_root().get_object ());
+			} catch (GLib.Error e) {
```
Comment (oleavr) [inline inject/inject.vala]: Should `printerr ()` here and then return.

File: inject/inject.vala [hunk 4]
```diff
@@ -59,7 +61,18 @@ namespace Frida.Inject {
...
 		}
 
-		application = new Application (target_pid, target_name, script_path, script_source, enable_jit, enable_development);
+		Json.Node parameters = new Json.Node (Json.NodeType.OBJECT);
+		parameters.set_object (new Json.Object ()); 
+		if (parameters_str != null && parameters_str != "") {
+			try {
+				var parser = new Json.Parser ();
+				parser.load_from_data (parameters_str, -1);
+				parameters.set_object (parser.get_root().get_object ());
```
Comment (oleavr) [inline inject/inject.vala]:
```suggestion
				parameters.set_object (parser.get_root ().get_object ());
```

File: inject/inject.vala [hunk 5]
```diff
@@ -117,6 +130,11 @@ namespace Frida.Inject {
 			construct;
 		}
 
+		public Json.Node? parameters {
```
Comment (oleavr) [inline inject/inject.vala]: Did you try making this non-nullable?

File: inject/inject.vala (hunk 2 above)
Comment (eugenekolo) [reply inject/inject.vala]: The empty string doesn't cause an error if you let it parse. "" is considered valid JSON in glib. 

Added a case for it, although I'm not sure it erroring is any better than just continuing on with an empty object.

File: inject/inject.vala [hunk 6]
```diff
@@ -59,7 +61,25 @@ namespace Frida.Inject {
...
+				printerr ("Parameters argument must be specified as JSON if present\n");
+				return 4;
+			}
+
+			try {
+				var parser = new Json.Parser ();
+				parser.load_from_data (parameters_str, -1);
+				parameters.set_object (parser.get_root ().get_object ());
+			} catch (GLib.Error e) {
+				printerr ("Failed to parse parameters argument as JSON\n");
```
Comment (oleavr) [inline inject/inject.vala]: Would be good to include the error message:

```suggestion
				printerr ("Failed to parse parameters argument as JSON: %s\n", e.message);
```

File: inject/inject.vala [hunk 7]
```diff
@@ -59,7 +61,25 @@ namespace Frida.Inject {
...
 		}
 
-		application = new Application (target_pid, target_name, script_path, script_source, enable_jit, enable_development);
+		var parameters = new Json.Node (Json.NodeType.OBJECT);
+		parameters.set_object (new Json.Object ()); 
+		if (parameters_str != null) {
+			if (parameters_str == "") {
+				printerr ("Parameters argument must be specified as JSON if present\n");
+				return 4;
+			}
```
Comment (oleavr) [inline inject/inject.vala]: Why not let it fail below – or is the error message unintuitive?

Comment (eugenekolo) [reply inject/inject.vala]: It seg faults.. so very unintuitive.

Comment (oleavr) [reply inject/inject.vala]: 🤣 Sounds like a bug in JSON-GLib. Anyway, let's work around it for now.

File: inject/inject.vala [hunk 8]
```diff
@@ -13,6 +14,7 @@ namespace Frida.Inject {
 		{ "pid", 'p', 0, OptionArg.INT, ref target_pid, null, "PID" },
 		{ "name", 'n', 0, OptionArg.STRING, ref target_name, null, "PID" },
 		{ "script", 's', 0, OptionArg.FILENAME, ref script_path, null, "JAVASCRIPT_FILENAME" },
+		{ "parameters",  's', 0, OptionArg.STRING, ref parameters_str, "Parameters as JSON, same as Gadget", "PARAMETERS_JSON" },
```
Comment (oleavr) [inline inject/inject.vala]: The short switch conflicts with `--scripts` / `-s`. Perhaps we should use `'P'`?

Comment (eugenekolo) [reply inject/inject.vala]: Changed to P

**Comment by oleavr** [comment]
Thanks, merged! 🤘

**Pull request merged**